circleguard/comparer.py

```diff
@@ -100,97 +103,126 @@ def _result(self, replay1, replay2):
-        mean = Comparer._compare_two_replays(replay1, replay2)
-        return StealResult(replay1, replay2, mean)
+
+        # perform preprocessing here as an optimization, so it is not repeated
+        # within different comparison algorithms. This will likely need to
+        # become more advanced if we add more (and different) algorithms.
+        xy1, xy2 = Comparer.interpolate(replay1, replay2)
+        xy1, xy2 = Comparer.clean(xy1, xy2)
+
+        # flip if one but not both has HR
+        if (Mod.HR in replay1.mods) ^ (Mod.HR in replay2.mods):
+            xy1[:, 1] = 384 - xy1[:, 1]
+
+        if Detect.STEAL_SIM & self.detect:
+            mean = Comparer.compute_similarity(xy1, xy2)
+            yield StealResultSim(replay1, replay2, mean)
+        if Detect.STEAL_CORR & self.detect:
+            correlation = Comparer.compute_correlation(xy1, xy2)
+            yield StealResultCorr(replay1, replay2, correlation)
+
 
     @staticmethod
-    def _compare_two_replays(replay1, replay2):
+    def compute_similarity(xy1, xy2):
         """
-        Calculates the average cursor distance between two
-        :class:`~.replay.Replay`\s, and the standard deviation of the distance.
+        Calculates the average distance between two sets of cursor position
+        data.
 
         Parameters
         ----------
-        replay1: :class:`~.replay.Replay`
-            The first replay to compare.
-        replay2: :class:`~.replay.Replay`
-            The second replay to compare.
+        replay1: ndarray
+            The first xy data to compare.
+        replay2: ndarray
+            The second xy data to compare.
 
         Returns
         -------
         float
-            The mean distance of the cursors of the two replays.
+            The mean distance between the two datasets.
+        """
+
+        # euclidean distance
+        distance = xy1 - xy2
+        distance = (distance ** 2).sum(axis=1) ** 0.5
+        return distance.mean()
+
+    @staticmethod
+    def compute_correlation(xy1, xy2, num_chunks=1):
+
+        xy1 = xy1.T
+        xy2 = xy2.T
+
+        # Sectioned into 20 chunks, used to ignore outliers (eg. long replay
+        # with breaks is copied, and the cheater cursordances during the break)
+        horizontal_length = xy1.shape[1] - xy1.shape[1] % num_chunks
+        xy1_sections = np.hsplit(xy1[:,:horizontal_length], num_chunks)
+        xy2_sections = np.hsplit(xy2[:,:horizontal_length], num_chunks)
+        correlations = []
+        for (xy1_section, xy2_section) in zip(xy1_sections, xy2_sections):
+            xy1_section -= np.mean(xy1_section)
+            xy2_section -= np.mean(xy2_section)
+            norm = np.std(xy1_section) * np.std(xy2_section) * xy1_section.size
+            cross_correlation_matrix = signal.correlate(xy1_section, xy2_section) / norm
+            # Pick the lag with the maximum correlation, this likely in
+            # most cases is 0 lag
```

Comment:
what does 0 lag mean? what is lag? I'm having a hard time parsing this comment in general.

Reply:
When you take the cross correlation of two identical signals that are out of phase, how much they are out of phase by is the lag of one signal wrt the other signal.

Because we're looking at two signals that are likely very similar, when we take the cross correlation we're given a matrix that measures the normalized cross correlation at various shifts in the time domain. In most cases, our two signals will be unshifted in the time domain relative to each other, and as such, in most cases grabbing the value with the highest correlation from this matrix is also correspondingly picking the the correlation associated with zero shift in the time domain.

So the latter part of the comment is really just to outline that in most cases, the maximum correlation happens we consider the two signals unshifted relative to each other.

Reply:
I see, thank you.

How large is the maximal time shift `signal.correlate` checks? Is it possible to bound it to `t ∈ [-2000, 2000]` ms or so, and save processing time? Because we can reasonably discard very large timeshifts as impossible to generate a high scoring replay (ie an FC)

Reply:
Not without essentially implementing a version of signal.correlate. I think you would rewrite that method, but use less zero padding in this step https://github.com/scipy/scipy/blob/v1.4.1/scipy/signal/signaltools.py#L240

Alternatively, I did some testing, and we could use

```python
cross_corr_matrix = signal.correlate(xy1_part, xy2_part, mode='valid', method='direct') / norm
```

which is dramatically faster than currently (about 5-6x). But if I'm interpreting this correlate, I think it kills the ability to detect time shifts.

You can test it with the shifted test case you created to see if that's the case.

Reply:
That speeds it up to within 25% or so of the current similarity method for me, but you're right that it doesn't detect time shifts - the one I tested, anyway. Oh well, looks like there's nothing to be gained here then, unless we were to take a deep dive into creating our own correlation for potentially little benefit.

Keeping `method="direct"` in the back of my mind as a potential fast-enough alternative to the current similarity though, if something were to ever go wrong with it.

Reply:
How did you benchmark it? I purely benchmarked the inner loop.

eg.

```python
        for (xy1_part, xy2_part) in zip(xy1_parts, xy2_parts):
            xy1_part -= np.mean(xy1_part)
            xy2_part -= np.mean(xy2_part)
            norm = np.std(xy1_part) * np.std(xy2_part) * xy1_part.size
            cross_corr_matrix = signal.correlate(xy1_part, xy2_part) / norm
            # Pick the lag with the maximum correlation, this likely in
            # most cases is 0 lag
            max_corr = np.max(cross_corr_matrix)
            correlations.append(max_corr)
```

done a thousand times.

Does the current method work against that time shifted replay? I'm actually still kinda unconvinced about the math that's (ab)used to detect time shifts with cross correlation.

Reply:
I benchmarked quite a bit outside that, but I'd be surprised if there was a (relative) time difference in moving the benchmarking towards the inside.

```

r1 = ReplayPath("/Users/tybug/Desktop/replays/stealing/stolen/stolen-time-shifted-1.osr")
r2 = ReplayPath("/Users/tybug/Desktop/replays/stealing/stolen/stolen-time-shifted-2.osr")


import time
t0 = time.time()
for _ in range(500):
    list(cg.steal_check([r1, r2], method=Detect.STEAL_CORR))
t1 = time.time()
print(t1 - t0)
``

The current method does work against the time shifted replay, yep! A convincing `0.9976412969068766`.

Reply:
Nice, I guess the zero padding is robust after the inputs are normalized.